nix/eval-machine-info.nix

```diff
@@ -11,104 +11,136 @@
-  nixpkgs = builtins.unsafeDiscardStringContext pkgs.path;
+  inherit (stage1Eval.config) nixpkgs;
+
+  pkgs = nixpkgs.legacyPackages.${system} or import nixpkgs { inherit system; };
+
+in
+rec {
+
+  network = (stage1Eval.extendModules {
+    modules = [
+      ({ config, ... }: {
+        options = {
+          network = {
+            enableRollback = lib.mkEnableOption "network wide rollback";
+            description = mkOption {
+              type = types.str;
+              description = "A description of the entire network.";
+              default = "";
+            };
+            nodesExtraArgs = mkOption {
+              description = "Extra inputs to be passed to every node.";
+              type = with types;attrsOf anything;
+              default = {};
+            };
+          };
+          resources = mkOption {
+            type = types.submoduleWith {
+              modules = [{
+                # so what is this trying to do?
+                sshKeyPairs = evalResources ./ssh-keypair.nix (lib.zipAttrs resourcesByType.sshKeyPairs or [ ]);
+                commandOutput = evalResources ./command-output.nix (lib.zipAttrs resourcesByType.commandOutput or [ ]);
+                machines = config.nodes;
+                _module.check = false;
+              }] ++ pluginResources;
+              specialArgs = {
+                inherit evalResources resourcesByType;
+                inherit (lib) zipAttrs;
+              };
+            };
+          };
+          # Compute the definitions of the machines.
+          nodes = mkOption {
+            type = types.attrsOf (types.submoduleWith {
+              specialArgs = {
+                inherit uuid deploymentName;
+                inherit (config) nodes resources;
+              } // config.network.nodesExtraArgs;
+              modules = (import "${nixpkgs}/nixos/modules/module-list.nix") ++
+                config.defaults ++
+                # Make NixOps's deployment.* options available.
+                pluginOptions ++
+                [
+                  ./options.nix
+                  ./resource.nix
+                  deploymentInfoModule
+                  ({ name, ... }: rec{
+                    _file = ./eval-machine-info.nix;
+                    key = _file;
+                    # Provide a default hostname and deployment target equal
+                    # to the attribute name of the machine in the model.
+                    networking.hostName = lib.mkOverride 900 name;
+                    deployment.targetHost = lib.mkOverride 900 name;
+                    environment.checkConfigurationOptions = lib.mkOverride 900 checkConfigurationOptions;
+                    nixpkgs.system = lib.mkDefault system;
+                  })
+                ];
+            });
+          };
+          defaults = mkOption {
+            type = with types;
+              # TODO: use types.raw once this PR is merged: https://github.com/NixOS/nixpkgs/pull/132448
```

Comment:
I think we'll want a type that doesn't prohibit merging, but returns a list or a `mkMerge` instead.

This option is actually somewhat redundant, as you can merge modules into all nodes by redeclaring `options.nodes`, but `defaults` is expected and convenient.